service/firewall/interception/dnslistener/module.go

```diff
@@ -0,0 +1,119 @@
+package dnslistener
+
+import (
+	"errors"
+	"net"
+	"sync/atomic"
+
+	"github.com/miekg/dns"
+	"github.com/safing/portmaster/base/log"
+	"github.com/safing/portmaster/service/integration"
+	"github.com/safing/portmaster/service/mgr"
+	"github.com/safing/portmaster/service/network/netutils"
+	"github.com/safing/portmaster/service/resolver"
+)
+
+var ResolverInfo = resolver.ResolverInfo{
+	Name:   "SystemdResolver",
+	Type:   "env",
+	Source: "System",
+}
+
+type DNSListener struct {
+	instance instance
+	mgr      *mgr.Manager
+
+	listener *Listener
+}
+
+// Manager returns the module manager.
+func (dl *DNSListener) Manager() *mgr.Manager {
+	return dl.mgr
+}
+
+// Start starts the module.
+func (dl *DNSListener) Start() error {
+	// Initialize dns event listener
+	var err error
+	dl.listener, err = newListener(dl)
+	if err != nil {
+		log.Errorf("failed to start dns listener: %s", err)
+	}
+
+	return nil
+}
+
+// Stop stops the module.
+func (dl *DNSListener) Stop() error {
+	if dl.listener != nil {
+		err := dl.listener.stop()
+		if err != nil {
+			log.Errorf("failed to close listener: %s", err)
+		}
+	}
+	return nil
+}
+
+// Flush flushes the buffer forcing all events to be processed.
+func (dl *DNSListener) Flush() error {
+	return dl.listener.flush()
+}
+
+func saveDomain(domain string, ips []net.IP, cnames map[string]string) {
+	for _, ip := range ips {
+		// Never save domain attributions for localhost IPs.
+		if netutils.GetIPScope(ip) == netutils.HostLocal {
+			continue
+		}
+		fqdn := dns.Fqdn(domain)
+
+		// Create new record for this IP.
+		record := resolver.ResolvedDomain{
+			Domain:            fqdn,
+			Resolver:          &ResolverInfo,
+			DNSRequestContext: &resolver.DNSRequestContext{},
+			Expires:           0,
+		}
+
+		for {
+			nextDomain, isCNAME := cnames[domain]
+			if !isCNAME {
+				break
+			}
```

Comment:
I just fixed a potential endless loop for CNAMES last week. :laughing: 
See https://github.com/safing/portmaster/commit/f4b96e1ce7ba95893c8dff929f6e0ef1e5b22809